src/jobservice/job/impl/gc/garbage_collection.go

```diff
@@ -186,6 +114,162 @@ func (gc *GarbageCollector) init(ctx job.Context, params job.Parameters) error {
+				continue
+			}
+		}
+		gc.logger.Infof("blob eligible for deletion: %s", blob.Digest)
+		gc.deleteSet = append(gc.deleteSet, blob)
+		// as table blob has no repository name, here needs to use the ArtifactTrash to fill it in.
+		if blob.IsManifest() {
+			for _, at := range arts {
+				if at.Digest == blob.Digest {
+					blob.Repositories = append(blob.Repositories, at.RepositoryName)
+				}
+			}
+			mfCt++
+		} else {
+			blobCt++
+		}
+	}
+	gc.logger.Infof("%d blobs and %d manifests eligible for deletion", blobCt, mfCt)
+	return nil
+}
+
+func (gc *GarbageCollector) sweep(ctx job.Context) error {
+	gc.logger = ctx.GetLogger()
+	for _, blob := range gc.deleteSet {
+		// set the status firstly, if the blob is updated by any HEAD/PUT request, it should be fail and skip.
+		blob.Status = blob_models.StatusDeleting
+		_, err := gc.blobMgr.UpdateBlobStatus(ctx.SystemContext(), blob)
+		if err != nil {
+			gc.logger.Errorf("failed to mark gc candidate deleting, skip: %s, %s", blob.Digest, blob.Status)
+			continue
+		}
+
+		if blob.IsManifest() {
+			for _, repo := range blob.Repositories {
+				// Harbor cannot know the existing tags in the backend from its database, so let the v2 DELETE manifest to remove all of them.
+				gc.logger.Infof("delete the manifest with registry v2 API: %s, %s, %s",
+					repo, blob.ContentType, blob.Digest)
+				if err := v2DeleteManifest(repo, blob.Digest); err != nil {
+					gc.logger.Errorf("failed to delete manifest with v2 API, %s, %s, %v", repo, blob.Digest, err)
+					if err := gc.markDeleteFailed(ctx, blob); err != nil {
+						return err
+					}
+					return errors.Wrapf(err, "failed to delete manifest with v2 API: %s, %s", repo, blob.Digest)
+				}
+				// for manifest, it has to delete the revisions folder of each repository
+				gc.logger.Infof("delete manifest from storage: %s", blob.Digest)
+				if err := ignoreNotFound(func() error {
+					return gc.registryCtlClient.DeleteManifest(repo, blob.Digest)
+				}); err != nil {
+					if err := gc.markDeleteFailed(ctx, blob); err != nil {
+						return err
+					}
+					return errors.Wrapf(err, "failed to remove manifest from storage: %s, %s", repo, blob.Digest)
+				}
+			}
+		}
+
+		// delete all of blobs, which include config, layer and manifest
+		gc.logger.Infof("delete blob from storage: %s", blob.Digest)
+		if err := ignoreNotFound(func() error {
+			return gc.registryCtlClient.DeleteBlob(blob.Digest)
+		}); err != nil {
+			if err := gc.markDeleteFailed(ctx, blob); err != nil {
+				return err
+			}
+			return errors.Wrapf(err, "failed to delete blob from storage: %s, %s", blob.Digest, blob.Status)
+
+		}
+
+		// remove the blob record
+		if err := ignoreNotFound(func() error {
+			return gc.blobMgr.Delete(ctx.SystemContext(), blob.ID)
+		}); err != nil {
+			if err := gc.markDeleteFailed(ctx, blob); err != nil {
+				return err
+			}
+			return errors.Wrapf(err, "failed to delete blob from database: %s, %s", blob.Digest, blob.Status)
```

Comment:
Does this mean that we will break the GC job when deleting the blob in DB was failed?

Reply:
yes, do you mean that we can skip it?

Reply:
Yes, I think we can skip this blob when delete it failed and to handle the next one.